### PR TITLE
Added audio web component + minor tweaks

### PR DESCRIPTION
**Pull request by roxburghm**
Hi,

Thanks for the code. Some modifications if you want to include them.

I have added audio_instructions_web component and modified app.js and demo.html accordingly.

audio instructions web allows utterance of "In 100m turn left on to Church Lane", in the speech codes -> text use a "!" to include " on to Street Name" in the instruction.
screen shot 2017-10-12 at 12 06 49

Fixed a "bug" in base component to allow classes to be sent via options to add additional classes to the component.

Added "grid.w" to override, clumsily, the components grid width to allow styling to match your desired layout.

Changed layout of demo.html to include more buttons (route overview).

Thanks

Matt

### CODE REVIEW

**Comment by fabrik42** [comment]
Hi!
Thank you for your PR, I will have a detailled look at it over the weekend. :)

File: README.md [hunk 1]
```diff
@@ -2,6 +2,14 @@
 
 ffwdme.js is a JavaScript toolkit that aims to bring interactive GPS driving directions to the mobile browser.
 
+## Fork Reason
```
Comment (fabrik42) [inline README.md]: No need to put this into the README, please remove. :)

File: src/components/arrival_time/index.js [hunk 2]
```diff
@@ -6,6 +6,14 @@ var ArrivalTime = BaseIcon.extend({
...
   defaultUnit: 'Uhr',
 
+    constructor: function(options) {
+
+      if (typeof options.defaultUnit !== 'undefined') this.defaultUnit = options.defaultUnit;
+
+        this.base(options);
+
+    },
+
```
Comment (fabrik42) [inline src/components/arrival_time/index.js:16]: Good find! 👍

But you should check if `options` are actually passed to the constructor, to prevent errors. Something like `options || options = {}` before working with them.

File: src/components/audio_instructions_web/_styles.scss [hunk 3]
```diff
@@ -0,0 +1,4 @@
+.ffwdme-components-audio-instructions {
```
Comment (fabrik42) [inline src/components/audio_instructions_web/_styles.scss]: Could you modify the namespace to `ffwdme-components-audio-instructions-web`?

File: static/demo.html [hunk 4]
```diff
@@ -23,6 +23,12 @@
 
         <script src="demo/credentials.js"></script>
         <script src="demo/app.js"></script>
+
+        <style>
+            .ffwdme-debug-wrapper>button {
```
Comment (fabrik42) [inline static/demo.html]: Please move this to https://github.com/ffwdme/ffwdme.js/blob/master/src/debug/debug.scss

File: src/components/audio_instructions_web/index.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+    if (destinationDistance && destinationDistance < this.forwardMeters && action != this.lastAction){
+      this.player.play('FINISH');
+      this.lastAction = action;
+      return;
+    }
+
+    if (!distance || !nextDirection) return;
+
+    var action = this.audioAction(distance, nextDirection.turnType);
```
Comment (fabrik42) [inline src/components/audio_instructions_web/index.js:49]: I think it might make sense to pass the whole direction, not only the street name. This way one could use more information in the TTS than just the street name.

File: src/components/audio_instructions_web/web_audio_player.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    if (!this.canPlayAudio) return false;
+
+    direction =  this.audioData[directionCode];
+    if (typeof direction === 'undefined') {
+        console.log('no phrase for ' + directionCode);
+        return false;
+    }
+
+    if (typeof street !== undefined) {
+        direction = direction.replace('!', 'on to ' + street);
```
Comment (fabrik42) [inline src/components/audio_instructions_web/web_audio_player.js]: I think we should use another marker that allows this class to do more interpolation in the future.
Curly brackets are very common, I think this would work here too: `in 100m turn hard right {{street}}`

File: src/components/base/index.js [hunk 7]
```diff
@@ -55,6 +60,14 @@ var Base = ffwdme.Class.extend({
 
     grid.x && el.addClass('ffwdme-grid-x' + grid.x);
     grid.y && el.addClass('ffwdme-grid-y' + grid.y);
+    if (grid.w) {
```
Comment (fabrik42) [inline src/components/base/index.js:63]: I don't understand the purpose of this. 
The CSS class is defined in the class definition, but also gets overwritten by the values in the initializer? (but with the same values at the moment)

**Review by fabrik42 — CHANGES_REQUESTED** [review]

**Comment by fabrik42** [comment]
Hey,
first of all, this is a very cool feature! Thank you for taking the time to craft a PR.
There are a couple open questions / minor changes that I requested. After they have been fixed, I would be very happy to merge it.

**Comment by roxburghm** [comment]
I'll get to work on it. Thanks.

File: README.md (hunk 1 above)
Comment (roxburghm) [reply README.md]: done.

File: src/components/audio_instructions_web/_styles.scss (hunk 3 above)
Comment (roxburghm) [reply src/components/audio_instructions_web/_styles.scss]: done

File: src/components/audio_instructions_web/index.js (hunk 5 above)
Comment (roxburghm) [reply src/components/audio_instructions_web/index.js:49]: done

File: src/components/audio_instructions_web/web_audio_player.js (hunk 6 above)
Comment (roxburghm) [reply src/components/audio_instructions_web/web_audio_player.js]: done - also remove "on to" hard coded in player.

File: src/components/base/index.js (hunk 7 above)
Comment (roxburghm) [reply src/components/base/index.js:63]: the class gets appended I think so it was throwing a formatting error when I had a w-5 and a w-8 because  it couldn't decide what was happening with width.

File: static/demo.html (hunk 4 above)
Comment (roxburghm) [reply static/demo.html]: Done

File: src/components/arrival_time/index.js (hunk 2 above)
Comment (roxburghm) [reply src/components/arrival_time/index.js:16]: Done

File: .gitignore [hunk 8]
```diff
@@ -1,7 +1,7 @@
 .DS_Store
 .sass-cache
 .awspublish-*
-build
+#build
```
Comment (fabrik42) [inline .gitignore]: Please revert this change. :)

File: package.json [hunk 9]
```diff
@@ -40,5 +40,9 @@
     "vinyl-source-stream": "~1.1.0",
     "watchify": "~3.6.0"
   },
-  "dependencies": {}
+  "dependencies": {
+    "core-util-is": "^1.0.2",
+    "util-deprecate": "^1.0.2",
+    "wrappy": "^1.0.2"
+  }
```
Comment (fabrik42) [inline package.json]: What are these dependencies?

File: static/demo/demo.css [hunk 10]
```diff
@@ -41,3 +41,7 @@ html, body {
 #playground.hidden {
   display: none;
 }
+
+.ffwdme-debug-wrapper>button {
+    margin-top: 100px;
+}
```
Comment (fabrik42) [inline static/demo/demo.css]: Please move this here: https://github.com/ffwdme/ffwdme.js/blob/master/src/debug/debug.scss#L35

Or remove :)

**Review by fabrik42 — CHANGES_REQUESTED** [review]
Some small changes are left to do and there appeared some new changes in .gitignore and package.json?

Adding the additional fixtures is fine for me if you give the second one a better naming (describe what happens during the ride) :)

**Comment by roxburghm** [comment]
Tweaked messy repository and removed ealing files as they shouldn't have been there.

File: src/debug/debug.scss [hunk 11]
```diff
@@ -85,4 +85,9 @@
   span{
     float:right;
   }
+
+  .ffwdme-debug-wrapper>button {
```
Comment (fabrik42) [inline src/debug/debug.scss]: You are already in the `ffwdme-debug-wrapper` scope. `>button` should be sufficient here.

Comment (roxburghm) [reply src/debug/debug.scss]: Corrected.

**Comment by roxburghm** [comment]
All changes should now be done and pushed.

**Review by fabrik42 — APPROVED** [review]

**Comment by fabrik42** [comment]
Thank you for your work and patience!

**Comment by roxburghm** [comment]
No problem at all, glad to help. Thanks for the package in the first place.
